src/control/lib/control/telemetry.go

```diff
@@ -0,0 +1,357 @@
+	for label := range m {
+		result = append(result, label)
+	}
+	sort.Strings(result)
+	return result
+}
+
+// Keys gets the sorted list of quantile keys.
+func (m QuantileMap) Keys() []float64 {
+	result := make([]float64, 0, len(m))
+	for q := range m {
+		result = append(result, q)
+	}
+	sort.Float64s(result)
+	return result
+}
+
+type (
+	// MetricsListReq is used to request the list of metrics.
+	MetricsListReq struct {
+		Host string // Host to query for telemetry data
+		Port uint32 // Port to use for collecting telemetry data
+	}
+
+	// MetricsListResp contains the list of available metrics.
+	MetricsListResp struct {
+		AvailableMetricSets []*MetricSet `json:"available_metric_sets"`
+	}
+)
+
+// MetricsList fetches the list of available metric types from the DAOS nodes.
+func MetricsList(ctx context.Context, req *MetricsListReq) (*MetricsListResp, error) {
+	if req == nil {
+		return nil, errors.New("nil request")
+	}
+
+	if req.Host == "" {
+		return nil, errors.New("host must be specified")
+	}
+
+	if req.Port == 0 {
+		return nil, errors.New("port must be specified")
+	}
+
+	scraped, err := scrapeMetrics(ctx, req.Host, req.Port)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to list metrics")
+	}
+
+	resp := new(MetricsListResp)
+
+	list := make([]*MetricSet, 0)
+	for _, name := range scraped.Keys() {
+		mf := scraped[name]
+		newMetric := &MetricSet{
+			Name:        name,
+			Description: mf.GetHelp(),
+			Type:        metricTypeFromPrometheus(mf.GetType()),
+		}
+		list = append(list, newMetric)
+	}
+
+	resp.AvailableMetricSets = list
+	return resp, nil
+}
+
+type (
+	// MetricsQueryReq is used to query telemetry values.
+	MetricsQueryReq struct {
+		Host        string   // host to query for telemetry data
+		Port        uint32   // port to use for collecting telemetry data
+		MetricNames []string // if empty, collects all metrics
+	}
+
+	// MetricsQueryResp contains the list of telemetry values per host.
+	MetricsQueryResp struct {
+		MetricSets []*MetricSet `json:"metric_sets"`
+	}
+)
+
+// MetricsQuery fetches the requested metrics values from the DAOS nodes.
+func MetricsQuery(ctx context.Context, req *MetricsQueryReq) (*MetricsQueryResp, error) {
+	if req == nil {
+		return nil, errors.New("nil request")
+	}
+
+	if req.Host == "" {
+		return nil, errors.New("host must be specified")
+	}
+
+	if req.Port == 0 {
+		return nil, errors.New("port must be specified")
+	}
+
+	scraped, err := scrapeMetrics(ctx, req.Host, req.Port)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to query metrics")
+	}
+
+	metricNames := req.MetricNames
+	if len(metricNames) == 0 {
+		// when caller doesn't specify metrics, we include all of them
+		metricNames = scraped.Keys()
+	}
+
+	return newMetricsQueryResp(scraped, metricNames)
+}
+
+func newMetricsQueryResp(scraped pbMetricMap, metricNames []string) (*MetricsQueryResp, error) {
+	resp := new(MetricsQueryResp)
+
+	list := make([]*MetricSet, 0)
```

Comment:
minor: you can avoid some background copy operations and memory allocations by specifying list capacity here as len(metricNames)

Reply:
Good catch, including that in my follow-on.